crates/ruff/src/rules/flake8_pyi/rules/str_or_repr_defined_in_stub.rs

```diff
@@ -0,0 +1,114 @@
+use rustpython_parser::ast;
+use rustpython_parser::ast::{Ranged, Stmt};
+
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_semantic::analyze::visibility::is_abstract;
+
+use crate::autofix::edits::delete_stmt;
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+/// ## What it does
+/// Checks for redundant definitions of `__str__` or `__repr__` in stubs.
+///
+/// ## Why is this bad?
+/// These definitions are redundant with `object.__str__` or `object.__repr__`.
+///
+/// ## Example
+/// ```python
+/// class Foo:
+///     def __repr__(self) -> str:
+///         ...
+/// ```
+#[violation]
+pub struct StrOrReprDefinedInStub {
+    name: String,
+}
+
+impl AlwaysAutofixableViolation for StrOrReprDefinedInStub {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let StrOrReprDefinedInStub { name } = self;
+        format!("Defining `{name}` in a stub is almost always redundant")
+    }
+
+    fn autofix_title(&self) -> String {
+        let StrOrReprDefinedInStub { name } = self;
+        format!("Remove definition of `{name}`")
+    }
+}
+
+/// PYI029
+pub(crate) fn str_or_repr_defined_in_stub(checker: &mut Checker, stmt: &Stmt) {
+    let Stmt::FunctionDef(ast::StmtFunctionDef {
```

Comment:
Should this rule also apply for `async` function definitions? 

If so, then the following could be useful (I thought we already had this, but I wasn't able to find it with a quick search). 

In `rustpython-ast`  define a new `AnyFunctionDef` node that is a union over sync and async function definitions

```rust
#[derive(Copy, Clone, Debug)]
pub enum AnyFunctionDefinition<'a> {
	FunctionDef(&'a StmtFunctionDef),
	AsyncFuncitonDef(&'a StmtAsyncFunctionDef)
}

impl AnyFunctionDefinition<'_> {
	pub fn body(&self) -> Suite {
		match self {
			Self::FunctionDef(StmtFunctionDef { body, .. }) | Self::AsyncFuncitonDef(StmtAsyncFunctionDef { body, .. }) => body
	}

	// other accessor methods

	pub fn into_async(self) -> Option<&StmtAsyncFunctionDef> {
		if let Self::AsyncFunctionDef(async) { Some(async) } else { None }
	}
}

impl AstNode for AnyFunctionDef<'_> {
	... 
}

impl<'a> From<&'a StmtFunctionDef> for AnyFunctionDefinition<'a> {
	fn from(value: &'a StmtFunctionDef) -> Self {
		Self::FunctionDef(value) 
	}
}

// Same for `Async`
```

Reply:
(I can take a look at this separate from this PR.)